tls/s2n_psk.c

```diff
@@ -0,0 +1,157 @@
+
+    return S2N_SUCCESS;
+}
+
+int s2n_psk_free(struct s2n_psk *psk)
+{
+    notnull_check(psk);
+
+    GUARD(s2n_free(&psk->early_secret));
+    GUARD(s2n_free(&psk->identity));
+    GUARD(s2n_free(&psk->secret));
+
+    return S2N_SUCCESS;
+}
+
+int s2n_psk_calculate_binder_hash(struct s2n_connection *conn, s2n_hash_algorithm hash_alg,
+        const struct s2n_blob *partial_client_hello, struct s2n_blob *output_binder_hash)
+{
+    notnull_check(partial_client_hello);
+    notnull_check(output_binder_hash);
+
+    /* Retrieve the current transcript.
+     * The current transcript will be empty unless this handshake included a HelloRetryRequest. */
+    struct s2n_hash_state current_hash_state = {0};
+    GUARD(s2n_handshake_get_hash_state(conn, hash_alg, &current_hash_state));
+
+    /* Copy the current transcript to avoid modifying the original. */
+    DEFER_CLEANUP(struct s2n_hash_state hash_copy, s2n_hash_free);
+    GUARD(s2n_hash_new(&hash_copy));
+    GUARD(s2n_hash_copy(&hash_copy, &current_hash_state));
+
+    /* Add the partial client hello to the transcript. */
+    GUARD(s2n_hash_update(&hash_copy, partial_client_hello->data, partial_client_hello->size));
+
+    /* Get the transcript digest */
+    GUARD(s2n_hash_digest(&hash_copy, output_binder_hash->data, output_binder_hash->size));
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_tls13_keys_init_with_psk(struct s2n_tls13_keys *keys, struct s2n_psk *psk)
+{
+    notnull_check(keys);
+
+    keys->hash_algorithm = psk->hash_alg;
+    GUARD(s2n_hash_hmac_alg(keys->hash_algorithm, &keys->hmac_algorithm));
+    GUARD(s2n_hash_digest_size(keys->hash_algorithm, &keys->size));
+    GUARD(s2n_blob_init(&keys->extract_secret, keys->extract_secret_bytes, keys->size));
+    GUARD(s2n_blob_init(&keys->derive_secret, keys->derive_secret_bytes, keys->size));
+    GUARD(s2n_hmac_new(&keys->hmac));
+
+    return S2N_SUCCESS;
+}
+
+int s2n_psk_calculate_binder(struct s2n_psk *psk, const struct s2n_blob *binder_hash,
+        struct s2n_blob *output_binder)
+{
+    notnull_check(psk);
+    notnull_check(binder_hash);
+    notnull_check(output_binder);
+
+    DEFER_CLEANUP(struct s2n_tls13_keys psk_keys, s2n_tls13_keys_free);
+    GUARD(s2n_tls13_keys_init_with_psk(&psk_keys, psk));
+    eq_check(binder_hash->size, psk_keys.size);
+    eq_check(output_binder->size, psk_keys.size);
+
+    /* Make sure the early secret is saved on the psk structure for later use */
+    GUARD(s2n_alloc(&psk->early_secret, psk_keys.size));
+    GUARD(s2n_blob_init(&psk_keys.extract_secret, psk->early_secret.data, psk_keys.size));
+
+    /* Derive the binder key */
+    GUARD(s2n_tls13_derive_binder_key_secret(&psk_keys, psk));
+    struct s2n_blob *binder_key = &psk_keys.derive_secret;
+
+    /* Expand the binder key into the finished key */
+    s2n_tls13_key_blob(finished_key, psk_keys.size);
+    GUARD(s2n_tls13_derive_finished_key(&psk_keys, binder_key, &finished_key));
```

Comment:
From our discussion offline: Could you add links to https://tools.ietf.org/html/rfc8446#section-4.2.11.2 and https://tools.ietf.org/html/rfc8446#section-4.4.4 to explain why we do this step? Thank-you.